server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/SingleDeviceViewOperator.java

```diff
@@ -0,0 +1,140 @@
+ * under the License.
+ */
+
+package org.apache.iotdb.db.mpp.execution.operator.process;
+
+import org.apache.iotdb.db.mpp.execution.operator.Operator;
+import org.apache.iotdb.db.mpp.execution.operator.OperatorContext;
+import org.apache.iotdb.tsfile.common.conf.TSFileDescriptor;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+import org.apache.iotdb.tsfile.read.common.block.TsBlock;
+import org.apache.iotdb.tsfile.read.common.block.column.BinaryColumnBuilder;
+import org.apache.iotdb.tsfile.read.common.block.column.Column;
+import org.apache.iotdb.tsfile.read.common.block.column.ColumnBuilder;
+import org.apache.iotdb.tsfile.read.common.block.column.NullColumn;
+import org.apache.iotdb.tsfile.read.common.block.column.RunLengthEncodedColumn;
+import org.apache.iotdb.tsfile.utils.Binary;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.util.List;
+
+/**
+ * The SingleDeviceViewOperator plays a similar role with DeviceViewOperator of adding a device
+ * column to current resultSet.
+ *
+ * <p>Different from DeviceViewOperator which merge the resultSet from different devices,
+ * SingleDeviceViewOperator only focuses on one single device, the goal of it is to add a device
+ * view. It's just a transition and won't change the way data flows.
+ */
+public class SingleDeviceViewOperator implements ProcessOperator {
+
+  private final OperatorContext operatorContext;
+  private final String device;
+  private final Operator deviceOperator;
+  // Used to fill columns and leave null columns which doesn't exist in some devices.
+  private final List<Integer> deviceColumnIndex;
+  // Column dataTypes that includes device column
+  private final List<TSDataType> dataTypes;
+
+  public SingleDeviceViewOperator(
+      OperatorContext operatorContext,
+      String device,
+      Operator deviceOperator,
+      List<Integer> deviceColumnIndex,
+      List<TSDataType> dataTypes) {
+    this.operatorContext = operatorContext;
+    this.device = device;
+    this.deviceOperator = deviceOperator;
+    this.deviceColumnIndex = deviceColumnIndex;
+    this.dataTypes = dataTypes;
+  }
+
+  @Override
+  public OperatorContext getOperatorContext() {
+    return operatorContext;
+  }
+
+  @Override
+  public ListenableFuture<?> isBlocked() {
+    ListenableFuture<?> blocked = deviceOperator.isBlocked();
+    if (!blocked.isDone()) {
+      return blocked;
+    }
+    return NOT_BLOCKED;
+  }
+
+  @Override
+  public TsBlock next() {
+    TsBlock tsBlock = deviceOperator.next();
+    if (tsBlock == null) {
+      return null;
+    }
+    // fill existing columns
+    Column[] newValueColumns = new Column[dataTypes.size()];
+    for (int i = 0; i < deviceColumnIndex.size(); i++) {
+      newValueColumns[deviceColumnIndex.get(i)] = tsBlock.getColumn(i);
+    }
+    // construct device column
+    ColumnBuilder deviceColumnBuilder = new BinaryColumnBuilder(null, 1);
+    deviceColumnBuilder.writeBinary(new Binary(device));
```

Comment:
this `new Binary(device)` should be a constant value, no need to new one in each next call

Reply:
actually, we can store a constant BinaryColumn directly, no need to build each time using ColumnBuilder